src/main/java/org/opensearch/security/support/StreamableRegistry.java

```diff
@@ -0,0 +1,123 @@
+    private final Map<Integer, Entry> idToEntryMap = new HashMap<>();
+
+    private StreamableRegistry() {
+        registerAllStreamables();
+    }
+
+    private static class Entry {
+        BaseWriteable.Writer<StreamOutput, Object> writer;
+        BaseWriteable.Reader<StreamInput, Object> reader;
+
+        Entry(BaseWriteable.Writer<StreamOutput, Object> writer, BaseWriteable.Reader<StreamInput, Object> reader) {
+            this.writer = writer;
+            this.reader = reader;
+        }
+    }
+
+    private BaseWriteable.Writer<StreamOutput, Object> getWriter(Class<?> clazz) {
+        if ( !classToIdMap.containsKey(clazz) ) {
+            throw new OpenSearchException(String.format("No writer registered for class %s", clazz.getName()));
+        }
+        return idToEntryMap.get(classToIdMap.get(clazz)).writer;
+    }
+
+    private BaseWriteable.Reader<StreamInput, Object> getReader(int id) {
+        if ( !idToEntryMap.containsKey(id) ) {
+            throw new OpenSearchException(String.format("No reader registered for id %s", id));
+        }
+        return idToEntryMap.get(id).reader;
+    }
+
+    private int getId(Class<?> clazz) {
+        if ( !classToIdMap.containsKey(clazz) ) {
+            throw new OpenSearchException(String.format("No writer registered for class %s", clazz.getName()));
+        }
+        return classToIdMap.get(clazz);
+    }
+
+    protected boolean isStreamable(Class<?> clazz) {
+        return classToIdMap.containsKey(clazz);
+    }
+
+    protected void writeTo(StreamOutput out, Object object) throws IOException {
+        out.writeByte((byte) getId(object.getClass()));
+        getWriter(object.getClass()).write(out, object);
+    }
+
+    protected Object readFrom(StreamInput in) throws IOException {
+        int id = in.readByte();
+        return getReader(id).read(in);
+    }
+
+    protected static StreamableRegistry getInstance() {
+        return instance;
+    }
+
+    protected void registerStreamable(Class<?> clazz, BaseWriteable.Writer<StreamOutput, Object> writer, BaseWriteable.Reader<StreamInput, Object> reader) {
+        Integer id = classToIdMap.size() + 1;
+        classToIdMap.put(clazz, id);
+        idToEntryMap.put(id, new Entry(writer, reader));
+    }
+
+    protected int getStreamableID(Class<?> clazz) {
+        if (!isStreamable(clazz)) {
+            throw new OpenSearchException(String.format("class %s is in streamable registry", clazz.getName()));
+        }  else {
+            return classToIdMap.get(clazz);
+        }
+    }
+
+    /**
+     * Register all streamables here. Register new streamables towards the end.
+     * Removing / reordering a registered streamable will change the typeIDs associated with the streamables
+     * causing a breaking change in the serialization format.
+     */
+    private void registerAllStreamables() {
+
+        // InetSocketAddress
+        this.registerStreamable(
+            InetSocketAddress.class,
```

Comment:
If I am understanding this correctly, a streamable is generally a class outside of this repository or another opensearch repo that the security plugin is dependent on that we would like to serialize and de-serialize on transport requests between nodes?

Reply:
Yes, this is the correct understanding. 

Original idea was to add this logic to core's StreamOutput/StreamInput classes. These classes recently underwent some refactoring - https://github.com/opensearch-project/OpenSearch/issues/5910. I came up with this implementation as it's decoupled from core and we can ensure that there are no typeID collisions.